File: Array&Hashing/C++/TwoSum.cpp
#include <map>
#include <iostream>
#include <vector>
using namespace std;

class Solution 
{
public:
    vector<int> twoSum(vector<int>& nums, int target) 
    {
        map<int, int> hash_map;

        vector<int> ret;

        for( int i = 0; i < nums.size(); i++ )
        {
            int diff = target - nums[i];

            if( hash_map.count(diff) )
            {
                ret = {i, hash_map[diff]};
            }

            hash_map[nums[i]] = i;
        }

        return ret;
    }
};

int main()
{
    vector<int> nums = {2,7,11,15};
    int target = 18;
    
    Solution s;
    
    vector<int> ret = s.twoSum(nums, target);

    for(int i=0; i< ret.size(); i++)
    {
        cout << ret[i] << endl;
    }
}

File: Array&Hashing/python/TwoSum.py
"""
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
You may assume that each input would have exactly one solution, and you may not use the same element twice.
You can return the answer in any order.

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
"""
#%%
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:        
        hash_map = {}
        
        for i, n in enumerate(nums):
            diff = target - n
            
            if diff in hash_map:
                return [hash_map[diff], i]     
            
            hash_map[n] = i

#%%
nums = [2,7,11,15]
s = Solution()

s.twoSum(nums, 9)
                

